### PR TITLE
HDDS-7145. Recon: Show last OM DB sync time on Overview page

### PR DESCRIPTION
**Pull request by smitajoshi12**
## What changes were proposed in this pull request?
Display ToolTip to show OM last updated Delta and Full timestamp.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7145

## How was this patch tested?
Manually Tested and reviewed by siyog and corrected review changes suggested by siyog.

### CODE REVIEW

**Comment by smengcl** [comment]
I have made some UI tweaks around the update time display. PTAL

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx [hunk 1]
```diff
@@ -27,6 +27,10 @@ import moment from 'moment';
 interface IAutoReloadPanelProps extends RouteComponentProps<object> {
   onReload: () => void;
   lastUpdated: number;
+  lastUpdatedOMDBDelta: number;
+  lastUpdatedOMDBFull: number;
+  lastUpdatedOMDBDeltaText: string;
+  lastUpdatedOMDBFullText: string;
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:33]: We can get rid of the variable here and just use hard-coded text right?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx [hunk 2]
```diff
@@ -38,21 +42,47 @@ class AutoReloadPanel extends React.Component<IAutoReloadPanelProps> {
...
         >
-          {moment(lastUpdated).format('LTS')}
+          {moment(lastUpdated).format('LT')}
         </Tooltip>
       );
+
+      const omDBDeltaFullToolTip = <span>
+          {lastUpdatedOMDBFullText}: {moment(lastUpdatedOMDBFull).fromNow()}, {moment(lastUpdatedOMDBFull).format('LT')}
+          <br/>
+          {lastUpdatedOMDBDeltaText}: {moment(lastUpdatedOMDBDelta).fromNow()}, {moment(lastUpdatedOMDBDelta).format('LT')}
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:59]: Here, we could just use hard-coded text label right? e.g.

```suggestion
          Full update: {moment(lastUpdatedOMDBFull).fromNow()}, {moment(lastUpdatedOMDBFull).format('LT')}
          <br/>
          Delta update: {lastUpdatedOMDBDeltaText}: {moment(lastUpdatedOMDBDelta).fromNow()}, {moment(lastUpdatedOMDBDelta).format('LT')}
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx [hunk 3]
```diff
@@ -27,6 +27,10 @@ import moment from 'moment';
 interface IAutoReloadPanelProps extends RouteComponentProps<object> {
   onReload: () => void;
   lastUpdated: number;
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:29]: Maybe we could do a small refactoring here, like you changed the text from "updated" to "refreshed" at the already existing timestamp on the page.
```suggestion
  lastRefreshed: number;
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx [hunk 4]
```diff
@@ -38,21 +42,47 @@ class AutoReloadPanel extends React.Component<IAutoReloadPanelProps> {
   };
 
   render() {
-    const {onReload, lastUpdated, isLoading} = this.props;
-    const lastUpdatedText = lastUpdated === 0 ? 'NA' :
+    const {onReload, lastUpdated, lastUpdatedOMDBDelta,lastUpdatedOMDBFull,isLoading,lastUpdatedOMDBDeltaText,lastUpdatedOMDBFullText} = this.props;
+    
+     const lastUpdatedText = lastUpdated === 0 || lastUpdated === undefined ? 'NA' :
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:47]: Here as well.
```suggestion
     const lastRefreshedText = lastRefreshed === 0 || lastRefreshed === undefined ? 'NA' :
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx [hunk 5]
```diff
@@ -38,21 +42,47 @@ class AutoReloadPanel extends React.Component<IAutoReloadPanelProps> {
...
         </Tooltip>
       );
+
+      const omDBDeltaFullToolTip = <span>
+          {lastUpdatedOMDBFullText}: {moment(lastUpdatedOMDBFull).fromNow()}, {moment(lastUpdatedOMDBFull).format('LT')}
+          <br/>
+          {lastUpdatedOMDBDeltaText}: {moment(lastUpdatedOMDBDelta).fromNow()}, {moment(lastUpdatedOMDBDelta).format('LT')}
+      </span>
+
+      const lastUpdatedDeltaToolTip = lastUpdatedOMDBDelta === 0 || lastUpdatedOMDBDelta === undefined || lastUpdatedOMDBFull === 0 || lastUpdatedOMDBFull === undefined ? 'NA' :
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:62]: This is not necessarily the last updated delta, so we can use the previously used name like this (just so the code can be more understandable later).
```suggestion
      const lastUpdatedDeltaFullToolTip = lastUpdatedOMDBDelta === 0 || lastUpdatedOMDBDelta === undefined || lastUpdatedOMDBFull === 0 || lastUpdatedOMDBFull === undefined ? 'NA' :
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx [hunk 6]
```diff
@@ -38,21 +42,47 @@ class AutoReloadPanel extends React.Component<IAutoReloadPanelProps> {
...
+      (
+        <Tooltip
+          placement='bottom' title={omDBDeltaFullToolTip}
+        >
+          {moment(lastUpdatedOMDBDelta).format('LT')}
+        </Tooltip>
+      );
+
+     const lastUpdatedDeltaText = lastUpdatedOMDBDelta === 0 || lastUpdatedOMDBDelta === undefined || lastUpdatedOMDBFull===0 || lastUpdatedOMDBFull === undefined ? '' :
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:71]: Same here.
```suggestion
     const lastUpdatedDeltaFullText = lastUpdatedOMDBDelta === 0 || lastUpdatedOMDBDelta === undefined || lastUpdatedOMDBFull === 0 || lastUpdatedOMDBFull === undefined ? '' :
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx [hunk 7]
```diff
@@ -75,7 +79,12 @@ export class Overview extends React.Component<Record<string, object>, IOverviewS
       buckets: 0,
       keys: 0,
       missingContainersCount: 0,
-      lastUpdated: 0
+      lastUpdated: 0,
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx:82]: Maybe change here as well to refer that this is the refresh timstamp.
```suggestion
      lastRefreshed: 0,
```

**Review by dombizita — COMMENTED** [review]
Thanks you for the patch @smitajoshi12! It looks good to me, I added a few inline comments that all are just naming related, so the code can be more understandable in the future. Let me now if some of them is not correct, in my understanding in this way it is more readable.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx (hunk 7 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx:82]: Done Changes as Suggested by You.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx (hunk 6 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:71]: Done Changes as Suggested by You.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx (hunk 5 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:62]: Done Changes as Suggested by You.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx (hunk 4 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:47]: Done Changes as Suggested by You.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx (hunk 3 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx:29]: Done Changes as Suggested by You.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx (hunk 1 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx]: Current Timestamp Can be from Full Snapshot or Delta Update so I have used  sort method in Descending order so If we do hardcoding of variables this logic will not apply so kept everything in separate variables.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx (hunk 2 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx]: Current Timestamp Can be from Full Snapshot or Delta Update so I have used sort method in Descending order so If we do hardcoding of variables this logic will not apply so kept everything in separate variables.

Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/autoReloadPanel/autoReloadPanel.tsx]: Sorry, I don't understand why we can't hard-code the texts here. If we hard-code here the texts (as we always want to display "`Full update`" or "`Delta update`") we don't even need the `lastUpdatedOMDBDeltaText` and `lastUpdatedOMDBFullText` variables on the `overview.tsx`, as we just add the `autoReloadPanel` to the overview page. We need two parameters for the `autoReloadPanel`, the two timestamps, the delta and full update times. So with the hard-code texts here we can also remove the text variables from the `overview.tsx` which would be easier. Let me know if I misunderstood something.

**Comment by smengcl** [comment]
@smitajoshi12 Please check your branch and do `git pull` before trying to `git push --force`. Now my previous commit to your branch has been discarded.

**Review by smengcl — COMMENTED** [review]
I have pushed two commits to fix some issues in this PR. Please take a look at what I have changed.

@smitajoshi12 You would want to do a `git pull` to retrieve changes I have made before making other changes (if any) to avoid discarding my changes again.

**Review by smengcl — APPROVED** [review]

**Comment by smengcl** [comment]
Thanks @smitajoshi12 for the patch. Thanks @dombizita for the review.

**Pull request merged**